### PR TITLE
RULEAPI-668: Take the sonar-security analyzer into account when collecting rule coverage in the coverage script

### PR DESCRIPTION
**Pull request by arseniy-sonar**
I hope you don't mind the first two commits that do just some refactoring without changing the functionality of the script (it does sort the fields in json now: for the ease of comparison later).
I validated that the script output did not change using `diff <(jq --sort-keys . ../covered_rules.json) <(jq --sort-keys . covered_rules.json)`.

The functional changes are in the last commit:

- adding `sonar-security` into the list of scanned repositories; 
- prefixing versions displayed in the coverage information with the repository name, since now the language rule can be implemented in multiple repositories.

### CODE REVIEW

File: rule_coverage/coverage.py [hunk 1]
```diff
@@ -59,26 +62,33 @@ def checkout(repo,version,batch_mode):
...
-    g=Git(repo)
-    g.checkout(version)
-    os.chdir(repo)
+    return Repo(repo)
+
+def scan_all_versions(repo):
+  git_repo = checkout_repo(repo)
+  os.chdir(repo)
+  g=Git(repo)
+  for tag in git_repo.tags:
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:70]: [education] why is scanning all versions necessary?

File: rule_coverage/coverage.py [hunk 2]
```diff
@@ -59,26 +62,33 @@ def checkout(repo,version,batch_mode):
...
+        print("checkout failed, resetting and cleaning")
+        g.reset('--hard',tag)
+        g.clean('-xfd')
+      dump_rules(repo,tag.name)
+  os.chdir('..')
+
+def scan_version(repo,version):
+  git_repo = checkout_repo(repo)
+  g=Git(repo)
+  g.checkout(version)
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:85]: [education] why we don't use `try...except` for this checkout?

File: rule_coverage/coverage.py [hunk 3]
```diff
@@ -59,26 +62,33 @@ def checkout(repo,version,batch_mode):
...
+  for tag in git_repo.tags:
+    if not '-' in tag.name:
+      print(f"{repo} {tag.name}")
+      try:
+        g.checkout(tag.name)
+      except Exception:
+        print("checkout failed, resetting and cleaning")
+        g.reset('--hard',tag)
+        g.clean('-xfd')
+      dump_rules(repo,tag.name)
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:79]: [education] shouldn't `dump_rules` be together with `checkout`?

File: rule_coverage/coverage.py [hunk 4]
```diff
@@ -49,7 +46,7 @@ def dump_rules(repo,version):
     sonarpedia = load_json(sp_file)
     path=str(sonarpedia_path)+'/'+sonarpedia['rules-metadata-path'].replace('\\','/')
     languages=sonarpedia['languages']
-    get_rules_json(path,languages,version)
+    get_rules_json(path,languages, repo + ' ' + version)
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:49]: [education] why was this change needed?

**Review by andrei-epure-sonarsource — COMMENTED** [review]
I have some questions

File: rule_coverage/coverage.py (hunk 1 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py:70]: In order to find the first version that implements the rule.
Also to find out if the rule was implemented in the past even if it is removed now. The latter point is the current behavior, which is arguably suboptimal. In the future, I think it is best to display the version first implemented the rule and the version last implemented it.

File: rule_coverage/coverage.py (hunk 4 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py:49]: With the addition of sonar-security, they can come from at least two repos: the sonar-language and sonar-security (at least for some languages).

File: rule_coverage/coverage.py [hunk 5]
```diff
@@ -6,12 +6,11 @@
...
-repos=['sonar-abap','sonar-cpp','sonar-cobol','sonar-dotnet','sonar-css','sonar-flex','slang-enterprise','sonar-java','SonarJS','sonar-php','sonar-pli','sonar-plsql','sonar-python','sonar-rpg','sonar-swift','sonar-tsql','sonar-vb','sonar-html','sonar-xml','sonar-kotlin', 'sonar-secrets']
-#repos=['sonar-php','sonar-pli','sonar-plsql','sonar-python','sonar-rpg','sonar-swift','sonar-tsql','sonar-vb','sonar-html','sonar-xml']
+repos=['sonar-abap','sonar-cpp','sonar-cobol','sonar-dotnet','sonar-css','sonar-flex','slang-enterprise','sonar-java','SonarJS','sonar-php','sonar-pli','sonar-plsql','sonar-python','sonar-rpg','sonar-swift','sonar-tsql','sonar-vb','sonar-html','sonar-xml','sonar-kotlin', 'sonar-secrets', 'sonar-security']
 
 def load_json(file):
   with open(file) as json_file:
-    return json.load(json_file) 
+    return json.load(json_file)
 
 def get_rules_json(path,languages,version):
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:15]: maybe use `repoAndVersion` rather than `version`? and the same for `dump_rule`, `store_rule`.

Because it may be confusing where it's the version or repoAndVersion. See `dump_rules` version and `get_rules_json` version

And maybe add here (or somewhere better) a comment to mention that some languages have rules in multiple repositories (sonar-security + sonar-lang)?

File: rule_coverage/coverage.py [hunk 6]
```diff
@@ -6,12 +6,11 @@
...
   print(f"Getting rules from {os.getcwd()} {path}")
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:16]: maybe log also `version` ?

File: rule_coverage/coverage.py (hunk 1 above)
Comment (andrei-epure-sonarsource) [reply rule_coverage/coverage.py:70]: do we have a ticket for this?

File: rule_coverage/coverage.py (hunk 2 above)
Comment (andrei-epure-sonarsource) [reply rule_coverage/coverage.py:85]: (given that we wrap the checkout in a `try` in `scan_all_versions`)

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
(minor comments)

File: rule_coverage/coverage.py (hunk 5 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py:15]: Sounds good. done.

File: rule_coverage/coverage.py (hunk 6 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py:16]: done.

File: rule_coverage/coverage.py [hunk 7]
```diff
@@ -6,52 +6,56 @@
...
+repos=['sonar-abap','sonar-cpp','sonar-cobol','sonar-dotnet','sonar-css','sonar-flex','slang-enterprise','sonar-java','SonarJS','sonar-php','sonar-pli','sonar-plsql','sonar-python','sonar-rpg','sonar-swift','sonar-tsql','sonar-vb','sonar-html','sonar-xml','sonar-kotlin', 'sonar-secrets', 'sonar-security']
 
 def load_json(file):
   with open(file) as json_file:
-    return json.load(json_file) 
+    return json.load(json_file)
 
-def get_rules_json(path,languages,version):
-  print(f"Getting rules from {os.getcwd()} {path}")
+# repoAndVersion uniquely identifies the analyzer and version implemeting
```
Comment (andrei-epure-sonarsource) [inline rule_coverage/coverage.py:15]:
```suggestion
# repoAndVersion uniquely identifies the analyzer and version implementing
```

File: rule_coverage/coverage.py (hunk 1 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py:73]: I created one: [RULEAPI-691](https://jira.sonarsource.com/browse/RULEAPI-691)

File: rule_coverage/coverage.py (hunk 3 above)
Comment (arseniy-sonar) [reply rule_coverage/coverage.py]: Yep, I've put them together now.
